Add support for ReadBatch/WriteBatch socket operations.

kormat opened the pull request:
N.B.: this only works with go 1.9+, and on linux.

Go 1.9 added support for reading and writing batches of packets from
sockets, via golang.org/x/net/ipv4. This patch adds support for this to
overlay/conn, and updates the BR to use it. Some performance testing
shows this gives an extra 4-500MBps performance to the BR.

Also:
- Remove gavv/monotime, now that we're on go1.9 which has monotonic time
  handling.
- Fix major bug in sockopt where supplied values were ignored.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/scionproto/scion/1348)
<!-- Reviewable:end -->

scrye commented:
Reviewed 10 of 10 files at r1.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[go/border/io.go, line 38 at r1](https://reviewable.io:443/reviews/scionproto/scion/1348#-L-kkcfkd4YpH6kN4kV5:-L-kkcflfs2yHJ1kaPVQ:be7ijnj) ([raw file](https://github.com/scionproto/scion/blob/42a9d63b1c89240ccccb58024f149a37ea32a7fb/go/border/io.go#L38)):*
> ```Go
> const (
> 	inputBufCnt       = 32
> 	inputBatchCnt     = 1 // Must be <= inputBufCnt
> ```

Why implement batching if batch size is set to `0`? Is the goal here to increase this in the future?

---

*[go/border/io.go, line 39 at r1](https://reviewable.io:443/reviews/scionproto/scion/1348#-L-kmcIDRMVsdWG1bQs3:-L-kmcIDRMVsdWG1bQs4:b-c5duke) ([raw file](https://github.com/scionproto/scion/blob/42a9d63b1c89240ccccb58024f149a37ea32a7fb/go/border/io.go#L39)):*
> ```Go
> 	inputBufCnt       = 32
> 	inputBatchCnt     = 1 // Must be <= inputBufCnt
> 	defInputLowBufCnt = 4
> ```

It's unclear to me what this represents, a comment would help here.

---

*[go/border/io.go, line 44 at r1](https://reviewable.io:443/reviews/scionproto/scion/1348#-L-klWrCDwCHmUPzmznR:-L-klWrCDwCHmUPzmznS:bwfiwjl) ([raw file](https://github.com/scionproto/scion/blob/42a9d63b1c89240ccccb58024f149a37ea32a7fb/go/border/io.go#L44)):*
> ```Go
> )
> 
> func (r *Router) posixInput(s *rctx.Sock, stop, stopped chan struct{}) {
> ```

This function has become very difficult to read at this point. The main loop is 100 lines long, and includes 4 loops, many branches and quite a lot of jumps.

---

*[go/border/io.go, line 80 at r1](https://reviewable.io:443/reviews/scionproto/scion/1348#-L-kmuWFb4BTyI66b4dj:-L-kmuWFb4BTyI66b4dk:b-s8oksp) ([raw file](https://github.com/scionproto/scion/blob/42a9d63b1c89240ccccb58024f149a37ea32a7fb/go/border/io.go#L80)):*
> ```Go
> 		default:
> 		}
> 		if len(pkts) < defInputLowBufCnt {
> ```

Suggestion: refactor this `if` and the next `for` to `pkts, msgs = buildNewMsgs(pkts, msgs, r.freePkts)` (or some other header that makes more sense). With a bit more effort, I think the helper function can be made to have no side-effects so that would help readability.

---

*[go/border/io.go, line 100 at r1](https://reviewable.io:443/reviews/scionproto/scion/1348#-L-koJE2TfAWtiZEytSn:-L-koJE2TfAWtiZEytSo:b11ywj2) ([raw file](https://github.com/scionproto/scion/blob/42a9d63b1c89240ccccb58024f149a37ea32a7fb/go/border/io.go#L100)):*
> ```Go
> 		}
> 		inputReads.Inc()
> 		toRead := len(pkts)
> ```

Defining a `min` allows `toRead = min(inputBatchCnt, len(pkts))` here.

---

*[go/border/io.go, line 106 at r1](https://reviewable.io:443/reviews/scionproto/scion/1348#-L-kqm2dGfSyidsGKH0-:-L-kqm2dGfSyidsGKH00:b-v1cpk2) ([raw file](https://github.com/scionproto/scion/blob/42a9d63b1c89240ccccb58024f149a37ea32a7fb/go/border/io.go#L106)):*
> ```Go
> 		var pktsRead int
> 		// Loop until a read succeeds, or a non-trivial error occurs
> 		for {
> ```

This can probably also be refactored using a helper `read` function to something like:
```
// Block until we read at least one message while ignoring ECONNREFUSED
pktsRead, _, err := readMsgs(...)
if err != nil {
    // metrics, logging
    continue Top
}
```

---

*[go/border/io.go, line 136 at r1](https://reviewable.io:443/reviews/scionproto/scion/1348#-L-krwZIf_2qY9odz33P:-L-krwZJsQ0FXG59sWH6:bb0ml3k) ([raw file](https://github.com/scionproto/scion/blob/42a9d63b1c89240ccccb58024f149a37ea32a7fb/go/border/io.go#L136)):*
> ```Go
> 		}
> 		// Loop over all read packets and set their metadata
> 		for i := 0; i < pktsRead; i++ {
> ```

This loop can also probably be moved to its own function, as it only uses the state within each packet, its corresponding message and socket `s`.

---

*[go/border/io.go, line 174 at r1](https://reviewable.io:443/reviews/scionproto/scion/1348#-L-kuDvwHB76l7d_YPbi:-L-kuDvxYVtMbCP0yk7k:b5kzo52) ([raw file](https://github.com/scionproto/scion/blob/42a9d63b1c89240ccccb58024f149a37ea32a7fb/go/border/io.go#L174)):*
> ```Go
> }
> 
> func (r *Router) posixOutput(s *rctx.Sock, _, stopped chan struct{}) {
> ```

Similarly to `posixInput`, this function can also benefit from a bit of refactoring.

---

*[go/border/io.go, line 192 at r1](https://reviewable.io:443/reviews/scionproto/scion/1348#-L-kx_2ya_qA6gf9u_8e:-L-kx_2ya_qA6gf9u_8f:b7hy7z0) ([raw file](https://github.com/scionproto/scion/blob/42a9d63b1c89240ccccb58024f149a37ea32a7fb/go/border/io.go#L192)):*
> ```Go
> 	outputWriteLatency := metrics.OutputWriteLatency.With(s.Labels)
> 
> 	var err error
> ```

Since there vars are only used within the `for`, moving them there improves clarity a bit regarding their lifetime.

---

*[go/border/io.go, line 197 at r1](https://reviewable.io:443/reviews/scionproto/scion/1348#-L-kxzw3P0S65HWwt1o-:-L-kxzw3P0S65HWwt1o0:b9rkins) ([raw file](https://github.com/scionproto/scion/blob/42a9d63b1c89240ccccb58024f149a37ea32a7fb/go/border/io.go#L197)):*
> ```Go
> 	var t float64
> 	for {
> 		if len(epkts) == 0 {
> ```

This can probably be moved to a helper function, together with the `for` loop below it.

---

*[go/border/io.go, line 220 at r1](https://reviewable.io:443/reviews/scionproto/scion/1348#-L-ky8mlnzNsaSHNNyCI:-L-ky8mlnzNsaSHNNyCJ:bza65k) ([raw file](https://github.com/scionproto/scion/blob/42a9d63b1c89240ccccb58024f149a37ea32a7fb/go/border/io.go#L220)):*
> ```Go
> 			}
> 		}
> 		toWrite := len(epkts)
> ```

`toWrite = min(outputBatchCnt, len(epkts))`

---

*[go/border/io.go, line 232 at r1](https://reviewable.io:443/reviews/scionproto/scion/1348#-L-kyTeSZ6JZsepaK5gW:-L-kyTeTgn_cxb7xVcS6:bnf85nh) ([raw file](https://github.com/scionproto/scion/blob/42a9d63b1c89240ccccb58024f149a37ea32a7fb/go/border/io.go#L232)):*
> ```Go
> 		t = time.Since(start).Seconds()
> 		bytes = 0
> 		for i := 0; i < pktsWritten; i++ {
> ```

This loop can also be refactored as a helper function that returns total `bytes` written.

---

*[go/lib/overlay/conn/conn.go, line 164 at r1](https://reviewable.io:443/reviews/scionproto/scion/1348#-L-kpFepsLtQzBvWW2Gn:-L-kpFepsLtQzBvWW2Go:bn7wism) ([raw file](https://github.com/scionproto/scion/blob/42a9d63b1c89240ccccb58024f149a37ea32a7fb/go/lib/overlay/conn/conn.go#L164)):*
> ```Go
> }
> 
> func (c *connUDPIPv4) ReadBatch(msgs []ipv4.Message, metas []ReadMeta) (int, *ReadMeta, error) {
> ```

Please add a docstring describing the return values, especially the second one as it's ignored in `border/io.go` and never set in this function (seems to always be `nil` for now?).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/scionproto/scion/1348)*
<!-- Sent from Reviewable.io -->

kormat commented:
Review status: 8 of 10 files reviewed at latest revision, 13 unresolved discussions.

---

*[go/border/io.go, line 38 at r1](https://reviewable.io:443/reviews/scionproto/scion/1348#-L-kkcfkd4YpH6kN4kV5:-L-l-QkJ5Yj8CvXFHQlE:b-s9ood4) ([raw file](https://github.com/scionproto/scion/blob/42a9d63b1c89240ccccb58024f149a37ea32a7fb/go/border/io.go#L38)):*
<details><summary><i>Previously, scrye (Sergiu Costea) wrote…</i></summary><blockquote>

Why implement batching if batch size is set to `0`? Is the goal here to increase this in the future?
</blockquote></details>

Oops, left-over from debugging. Fixed

---

*[go/border/io.go, line 39 at r1](https://reviewable.io:443/reviews/scionproto/scion/1348#-L-kmcIDRMVsdWG1bQs3:-L-l043FdV8ohlqzkYst:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/42a9d63b1c89240ccccb58024f149a37ea32a7fb/go/border/io.go#L39)):*
<details><summary><i>Previously, scrye (Sergiu Costea) wrote…</i></summary><blockquote>

It's unclear to me what this represents, a comment would help here.
</blockquote></details>

Done.

---

*[go/border/io.go, line 44 at r1](https://reviewable.io:443/reviews/scionproto/scion/1348#-L-klWrCDwCHmUPzmznR:-L-lFSY2djamDUBS_xhn:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/42a9d63b1c89240ccccb58024f149a37ea32a7fb/go/border/io.go#L44)):*
<details><summary><i>Previously, scrye (Sergiu Costea) wrote…</i></summary><blockquote>

This function has become very difficult to read at this point. The main loop is 100 lines long, and includes 4 loops, many branches and quite a lot of jumps.
</blockquote></details>

Done.

---

*[go/border/io.go, line 80 at r1](https://reviewable.io:443/reviews/scionproto/scion/1348#-L-kmuWFb4BTyI66b4dj:-L-l34kGklVA0xYhAnQA:b7he8jv) ([raw file](https://github.com/scionproto/scion/blob/42a9d63b1c89240ccccb58024f149a37ea32a7fb/go/border/io.go#L80)):*
<details><summary><i>Previously, scrye (Sergiu Costea) wrote…</i></summary><blockquote>

Suggestion: refactor this `if` and the next `for` to `pkts, msgs = buildNewMsgs(pkts, msgs, r.freePkts)` (or some other header that makes more sense). With a bit more effort, I think the helper function can be made to have no side-effects so that would help readability.
</blockquote></details>

I don't know what you mean by no side-effects, but i've pulled it out to a method.

---

*[go/border/io.go, line 100 at r1](https://reviewable.io:443/reviews/scionproto/scion/1348#-L-koJE2TfAWtiZEytSn:-L-l3VaJIT7n6pvOuHVe:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/42a9d63b1c89240ccccb58024f149a37ea32a7fb/go/border/io.go#L100)):*
<details><summary><i>Previously, scrye (Sergiu Costea) wrote…</i></summary><blockquote>

Defining a `min` allows `toRead = min(inputBatchCnt, len(pkts))` here.
</blockquote></details>

Done.

---

*[go/border/io.go, line 106 at r1](https://reviewable.io:443/reviews/scionproto/scion/1348#-L-kqm2dGfSyidsGKH0-:-L-l5462Cw9wJvZ3KjUr:b-dtuhd) ([raw file](https://github.com/scionproto/scion/blob/42a9d63b1c89240ccccb58024f149a37ea32a7fb/go/border/io.go#L106)):*
<details><summary><i>Previously, scrye (Sergiu Costea) wrote…</i></summary><blockquote>

This can probably also be refactored using a helper `read` function to something like:
```
// Block until we read at least one message while ignoring ECONNREFUSED
pktsRead, _, err := readMsgs(...)
if err != nil {
    // metrics, logging
    continue Top
}
```
</blockquote></details>

This will suck a bit when we add metrics for things like connrefused errors, but done for now.

---

*[go/border/io.go, line 136 at r1](https://reviewable.io:443/reviews/scionproto/scion/1348#-L-krwZIf_2qY9odz33P:-L-l5pr_mN5TwCq7pYTP:b-5yziw2) ([raw file](https://github.com/scionproto/scion/blob/42a9d63b1c89240ccccb58024f149a37ea32a7fb/go/border/io.go#L136)):*
<details><summary><i>Previously, scrye (Sergiu Costea) wrote…</i></summary><blockquote>

This loop can also probably be moved to its own function, as it only uses the state within each packet, its corresponding message and socket `s`.
</blockquote></details>

It uses:
- `pkts`
- `msgs`
- `readMetas`
- `s` (for `{Dir,Ifids,LocIdx}`)
- `free` closure
- 4 metrics
- `dst`

That's a lot of parameters for a function...

---

*[go/border/io.go, line 174 at r1](https://reviewable.io:443/reviews/scionproto/scion/1348#-L-kuDvwHB76l7d_YPbi:-L-lAcrrd0yapcIAafBC:bl2b8yf) ([raw file](https://github.com/scionproto/scion/blob/42a9d63b1c89240ccccb58024f149a37ea32a7fb/go/border/io.go#L174)):*
<details><summary><i>Previously, scrye (Sergiu Costea) wrote…</i></summary><blockquote>

Similarly to `posixInput`, this function can also benefit from a bit of refactoring.
</blockquote></details>

I've done the same as you suggested, moving the pkt & msg getting/setting out to a helper, which i think helps sufficiently. Thoughts?

---

*[go/border/io.go, line 192 at r1](https://reviewable.io:443/reviews/scionproto/scion/1348#-L-kx_2ya_qA6gf9u_8e:-L-lAop5v2_GDFvo53cO:b791i0k) ([raw file](https://github.com/scionproto/scion/blob/42a9d63b1c89240ccccb58024f149a37ea32a7fb/go/border/io.go#L192)):*
<details><summary><i>Previously, scrye (Sergiu Costea) wrote…</i></summary><blockquote>

Since there vars are only used within the `for`, moving them there improves clarity a bit regarding their lifetime.
</blockquote></details>

They were originally outside as `goto` disallowed them to be declared between `goto End` and `End`, but yeah, i can move them. It all looks a bit awkward in any case.

---

*[go/border/io.go, line 197 at r1](https://reviewable.io:443/reviews/scionproto/scion/1348#-L-kxzw3P0S65HWwt1o-:-L-lAkyAyYyiJTXdkX55:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/42a9d63b1c89240ccccb58024f149a37ea32a7fb/go/border/io.go#L197)):*
<details><summary><i>Previously, scrye (Sergiu Costea) wrote…</i></summary><blockquote>

This can probably be moved to a helper function, together with the `for` loop below it.
</blockquote></details>

Done.

---

*[go/border/io.go, line 220 at r1](https://reviewable.io:443/reviews/scionproto/scion/1348#-L-ky8mlnzNsaSHNNyCI:-L-lFPLpQbqmx48EPPgC:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/42a9d63b1c89240ccccb58024f149a37ea32a7fb/go/border/io.go#L220)):*
<details><summary><i>Previously, scrye (Sergiu Costea) wrote…</i></summary><blockquote>

`toWrite = min(outputBatchCnt, len(epkts))`
</blockquote></details>

Done.

---

*[go/border/io.go, line 232 at r1](https://reviewable.io:443/reviews/scionproto/scion/1348#-L-kyTeSZ6JZsepaK5gW:-L-lCmroM9_pjjjiSVzA:b-iksvsa) ([raw file](https://github.com/scionproto/scion/blob/42a9d63b1c89240ccccb58024f149a37ea32a7fb/go/border/io.go#L232)):*
<details><summary><i>Previously, scrye (Sergiu Costea) wrote…</i></summary><blockquote>

This loop can also be refactored as a helper function that returns total `bytes` written.
</blockquote></details>

Left this for now, as i think `posixOutput` is already significantly improved.

---

*[go/lib/overlay/conn/conn.go, line 164 at r1](https://reviewable.io:443/reviews/scionproto/scion/1348#-L-kpFepsLtQzBvWW2Gn:-L-l-yEMwYjTTfNgW01H:b-br0p0q) ([raw file](https://github.com/scionproto/scion/blob/42a9d63b1c89240ccccb58024f149a37ea32a7fb/go/lib/overlay/conn/conn.go#L164)):*
<details><summary><i>Previously, scrye (Sergiu Costea) wrote…</i></summary><blockquote>

Please add a docstring describing the return values, especially the second one as it's ignored in `border/io.go` and never set in this function (seems to always be `nil` for now?).
</blockquote></details>

Ah, obsolete, removed, and docstringed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/scionproto/scion/1348)*
<!-- Sent from Reviewable.io -->

scrye commented:
Review status: 8 of 10 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[go/border/io.go, line 80 at r1](https://reviewable.io:443/reviews/scionproto/scion/1348#-L-kmuWFb4BTyI66b4dj:-L-lJGXWpEjouNt_5UIV:b-iqp4ux) ([raw file](https://github.com/scionproto/scion/blob/42a9d63b1c89240ccccb58024f149a37ea32a7fb/go/border/io.go#L80)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

I don't know what you mean by no side-effects, but i've pulled it out to a method.
</blockquote></details>

Leftovers from functional programming days -- functions that take some inputs, produce some outputs and change nothing else. They're a nice thing to have because it makes code very easy to follow.

---

*[go/border/io.go, line 174 at r1](https://reviewable.io:443/reviews/scionproto/scion/1348#-L-kuDvwHB76l7d_YPbi:-L-lLOHacRaYvJi9AGCJ:bo4jkzf) ([raw file](https://github.com/scionproto/scion/blob/42a9d63b1c89240ccccb58024f149a37ea32a7fb/go/border/io.go#L174)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

I've done the same as you suggested, moving the pkt & msg getting/setting out to a helper, which i think helps sufficiently. Thoughts?
</blockquote></details>

Yes, it should be enough -- everything looks pretty clear now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/scionproto/scion/1348)*
<!-- Sent from Reviewable.io -->

scrye commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/scionproto/scion/1348)*
<!-- Sent from Reviewable.io -->

scrye commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/scionproto/scion/1348#-:-L-lMK8S98WxeWW1Vg3y:bnfp4nl)*
<!-- Sent from Reviewable.io -->
